aws/request/request.go

```diff
@@ -562,30 +561,39 @@ func AddToUserAgent(r *Request, s string) {
 	r.HTTPRequest.Header.Set("User-Agent", s)
 }
 
-func shouldRetryCancel(r *Request) bool {
-	awsErr, ok := r.Error.(awserr.Error)
-	timeoutErr := false
-	errStr := r.Error.Error()
-	if ok {
-		if awsErr.Code() == CanceledErrorCode {
+type temporary interface {
+	Temporary() bool
+}
+
+func shouldRetryCancel(err error) bool {
+	switch err := err.(type) {
+	case awserr.Error:
+		if err.Code() == CanceledErrorCode {
 			return false
 		}
-		err := awsErr.OrigErr()
-		netErr, netOK := err.(net.Error)
-		timeoutErr = netOK && netErr.Temporary()
-		if urlErr, ok := err.(*url.Error); !timeoutErr && ok {
-			errStr = urlErr.Err.Error()
+		return shouldRetryCancel(err.OrigErr())
+	case temporary:
+		// If the error is temporary, we want to
+		// allow continuation of the retry process.
+		return err.Temporary()
+	case *url.Error:
+		// *url.Error only implements Temporary
+		// after golang 1.6
+		// but since url.Error only wraps
+		// the error:
+		return shouldRetryCancel(err.Err)
+	case nil:
+		return true
```

Comment:
Is it possible for shouldRetryCanceled to be called with a nil error? If so, did the request actually fail? I'm not sure returning true instructing the SDK a request should be retried if there was no error is correct.

Reply:
Yes, well during unit tests it was. A nil error was wrapped. This caused a panic. So I thought since we don't know what exactly the error is : retry.
I figured I'd not fix these tests and handle the nil case to avoid a panic or an if in the `default:` section.

Reply:
This is the test case that made me leave the nil case: 

https://github.com/aws/aws-sdk-go/blob/fix_should_retry_cancel/aws/request/request_test.go#L82-L111

In those cases, somehow a nil error is wrapped.

Reply:
Thanks, that makes sense. Yeah adding r.Error is needed in the Send handler if the `StatusCode` is not a `200`.  This is done by the SDK's built in Send handler.

Reply:
Okay, adding this now.

Reply:
Thanks for looking into this. It looks like the SDK's test status is a little fragile. You're correct that the mocked send handler should not be setting the RequestFailure. This error will be set via the UnmarshalMeta handler after the request is sent.

I think the reason the panic in `shouldRetryCancel` is occurring is because the `return shouldRetryCancel(err.OrigErr())` on line request.go:574 of the PR. It is possible for `err.OrigErr()` to return nil, if there was no underlying error. In your pr the `nil` check case of `shouldRetryCancel` would effectively be catching this. 

We can keep the `nil` case, but it would be helpful to add docs to this case stating, that a nested error could be nil within a parent error.

Reply:
Gotcha, adding a doc 🙂